lib/ash_authentication/add_ons/audit_log/auditor.ex

```diff
@@ -0,0 +1,170 @@
+  @moduledoc """
+  Provides common audit logging behaviour for Ash actions.
+  """
+
+  alias __MODULE__
+  alias Spark.Dsl.Extension
+  require Logger
+
+  @type input :: Ash.ActionInput.t() | Ash.Changeset.t() | Ash.Query.t()
+  @type result :: {:ok, Ash.Resource.record()} | {:ok, [Ash.Resource.record()]} | {:error, any()}
+
+  defmodule Change do
+    @moduledoc "Implements the `Ash.Resource.Change` behaviour for audit logging"
+    use Ash.Resource.Change
+
+    @doc false
+    @impl true
+    def change(changeset, opts, context) do
+      tracked_actions = Auditor.get_tracked_actions(changeset.resource, opts[:strategy])
+
+      if changeset.action.name in tracked_actions do
+        Ash.Changeset.after_transaction(
+          changeset,
+          &Auditor.after_transaction(&1, &2, opts[:strategy], context)
+        )
+      else
+        changeset
+      end
+    end
+
+    @doc false
+    @impl true
+    def atomic(changeset, opts, context), do: {:ok, change(changeset, opts, context)}
+  end
+
+  defmodule Preparation do
+    @moduledoc "Implements the `Ash.Resource.Preparation` behaviour for audit logging"
+    use Ash.Resource.Preparation
+
+    @doc false
+    @impl true
+    def prepare(query, opts, context) when is_struct(query, Ash.Query) do
+      tracked_actions = Auditor.get_tracked_actions(query.resource, opts[:strategy])
+
+      if query.action.name in tracked_actions do
+        Ash.Query.after_transaction(
+          query,
+          &Auditor.after_transaction(&1, &2, opts[:strategy], context)
+        )
+      else
+        query
+      end
+    end
+
+    def prepare(input, opts, context) when is_struct(input, Ash.ActionInput) do
+      tracked_actions = Auditor.get_tracked_actions(input.resource, opts[:strategy])
+
+      if input.action.name in tracked_actions do
+        Ash.ActionInput.after_transaction(
+          input,
+          &Auditor.after_transaction(&1, &2, opts[:strategy], context)
+        )
+      else
+        input
+      end
+    end
+
+    @doc false
+    @impl true
+    def supports(_opts), do: [Ash.Query, Ash.ActionInput]
+  end
+
+  @doc false
+  @spec get_tracked_actions(Ash.Resource.t(), atom) :: [atom]
+  def get_tracked_actions(resource, strategy_name) do
+    Extension.get_persisted(resource, {:audit_log, strategy_name, :actions}) || []
+  end
+
+  @doc false
+  @spec after_transaction(input, result, atom, map) :: result
+  def after_transaction(input, result, strategy_name, context) do
+    audit_strategy = AshAuthentication.Info.strategy!(input.resource, strategy_name)
+
+    action_strategy =
+      AshAuthentication.Info.strategy_for_action!(input.resource, input.action.name)
+
+    status =
+      case result do
+        ok when elem(ok, 0) == :ok -> :success
+        error when elem(error, 0) == :error -> :failure
+        :error -> :failure
+        {ok, _} when elem(ok, 0) == :ok -> :success
+        {error, _} when elem(error, 0) == :error -> :failure
+        _ -> :unknown
```

Comment:
I think this could be written more ideomatic and I'm not sure if these lines

```elixir
        {ok, _} when elem(ok, 0) == :ok -> :success
        {error, _} when elem(error, 0) == :error -> :failure
```

are necessary? I can't see those in the type definition https://github.com/ash-project/ash/blob/b107cb0d4b59e9e69212ad6c5e129051c5620688/lib/ash/subject.ex#L43-L43

Reply:
You were right. This is definitely a bug. https://github.com/ash-project/ash/pull/2354